src/settings.cpp

```diff
@@ -495,50 +494,28 @@ static void Write_ValidateSetting(void *ptr, const SettingDesc *sd, int32 val)
 	WriteValue(ptr, sd->save.conv, (int64)val);
 }
 
-/**
- * Set the string value of a setting.
- * @param ptr Pointer to the storage location (might be a pointer to a pointer).
- * @param sld Pointer to the information for the conversions and limitations to apply.
- * @param p   The string to save.
- */
-static void Write_ValidateString(void *ptr, const SaveLoad *sld, const char *p)
-{
-	switch (GetVarMemType(sld->conv)) {
-		case SLE_VAR_STRB:
-		case SLE_VAR_STRBQ:
-			if (p != nullptr) {
-				char *begin = (char*)ptr;
-				char *end = begin + sld->length - 1;
-				strecpy(begin, p, end);
-				str_validate(begin, end, SVS_NONE);
-			}
-			break;
-
-		case SLE_VAR_STR:
-		case SLE_VAR_STRQ:
-			free(*(char**)ptr);
-			*(char**)ptr = p == nullptr ? nullptr : stredup(p);
-			break;
-
-		default: NOT_REACHED();
-	}
-}
-
 /**
  * Set the string value of a setting.
  * @param ptr Pointer to the std::string.
- * @param sld Pointer to the information for the conversions and limitations to apply.
+ * @param sd  Pointer to the information for the conversions and limitations to apply.
  * @param p   The string to save.
  */
-static void Write_ValidateStdString(void *ptr, const SaveLoad *sld, const char *p)
+static void Write_ValidateStdString(void *ptr, const SettingDesc *sd, const char *p)
 {
 	std::string *dst = reinterpret_cast<std::string *>(ptr);
 
-	switch (GetVarMemType(sld->conv)) {
+	switch (GetVarMemType(sd->save.conv)) {
 		case SLE_VAR_STR:
 		case SLE_VAR_STRQ:
 			if (p != nullptr) {
 				dst->assign(p);
+				if (sd->desc.max != 0 && dst->size() >= sd->desc.max) {
+					/* In case a maximum length is imposed by the setting, the length
+					 * includes the '\0' termination for network transfer purposes.
+					 * Also ensure the string is valid after chopping of some bytes. */
+					dst->resize(sd->desc.max - 1);
+					str_validate(*dst, SVS_NONE);
```

Comment:
The std::string str_validate does not change the passed string, but it constructs a new one. So as it is now, it is broken.